File: XT.js

// Check if a number is a power of two
function is_power_of_two(n) {
    return n !== 0 && (n & (n - 1)) === 0;
}

// Represent a power of two with dots
function represent_power_of_two(n) {
    let count = 0;
    while (n > 1) {
        n = Math.floor(n / 2);
        count += 1;
    }
    return '.'.repeat(count);
}

// Find all prime numbers up to a certain number (Sieve of Eratosthenes)
function find_primes_up_to(n) {
    let sieve = new Array(n + 1).fill(true);
    let p = 2;
    while (p * p <= n) {
        if (sieve[p]) {
            for (let i = p * p; i <= n; i += p) {
                sieve[i] = false;
            }
        }
        p += 1;
    }
    let primes = [];
    for (let i = 2; i <= n; i++) {
        if (sieve[i]) {
            primes.push(i);
        }
    }
    return primes;
}

// Represent a prime number based on its index in the list of primes
function represent_prime(prime, primes) {
    let index = primes.indexOf(prime) + 1;
    if (index === 1) {
        return '.';
    }
    return '(' + represent_number(index, primes) + ')';
}

// Represent a number as a power of two, a prime, or decomposed into prime factors
function represent_number(n, primes) {
    if (is_power_of_two(n)) {
        return represent_power_of_two(n);
    } else if (primes.includes(n)) {
        return represent_prime(n, primes);
    } else {
        let factors = decompose_into_primes(n, primes);
        return factors.map(factor => represent_prime(factor, primes)).join('');
    }
}

// Decompose a number into its prime factors
function decompose_into_primes(n, primes) {
    let factors = [];
    for (let i = 0; primes[i] * primes[i] <= n; i++) {
        while (n % primes[i] === 0) {
            factors.push(primes[i]);
            n /= primes[i];
        }
    }
    if (n > 1) {
        factors.push(n);
    }
    return factors;
}

File: CLI-scripts/XT.py
import time

def is_power_of_two(n):
    return n != 0 and ((n & (n - 1)) == 0)

def represent_power_of_two(n):
    count = 0
    while n > 1:
        n //= 2
        count += 1
    return '.' * count

def find_primes_up_to(n):
    sieve = [True] * (n + 1)
    p = 2
    while p * p <= n:
        if sieve[p]:
            for i in range(p * p, n + 1, p):
                sieve[i] = False
        p += 1
    return [p for p in range(2, n+1) if sieve[p]]

def represent_prime(prime, primes):
    index = primes.index(prime) + 1
    if index == 1:
        return '.'
    return '(' + represent_number(index, primes) + ')'

def represent_number(n, primes):
    if is_power_of_two(n):
        return represent_power_of_two(n)
    elif n in primes:
        return represent_prime(n, primes)
    else:
        factors = decompose_into_primes(n)
        return ''.join(represent_prime(factor, primes) for factor in factors)

def decompose_into_primes(n):
    i = 2
    factors = []
    while i * i <= n:
        if n % i:
            i += 1
        else:
            n //= i
            factors.append(i)
    if n > 1:
        factors.append(n)
    return factors

def formatted_time(time_ns):
    if time_ns < 1e3:
        return f"{time_ns} nanoseconds"
    elif time_ns < 1e6:
        return f"{time_ns / 1e3} microseconds"
    elif time_ns < 1e9:
        return f"{time_ns / 1e6} milliseconds"
    else:
        return f"{time_ns / 1e9} seconds"

# Main process
number = int(input("Enter a number: "))
start_time = time.time_ns()
maximum_number_for_primes = max(100, number * 2)
primes_up_to_maximum = find_primes_up_to(maximum_number_for_primes)
representation = represent_number(number, primes_up_to_maximum)
end_time = time.time_ns()

print(f"Representation of the number {number}: {representation}")
print(f"Time elapsed: {formatted_time(end_time - start_time)}")
